File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.js
function solution(board) {
    const dy = [-1, 1, 0, 0];
    const dx = [0, 0, -1, 1];

    const N = board.length;

    const cost = Array.from({ length: N }, () =>
        Array.from({ length: N }, () => Array(4).fill(Number.MAX_SAFE_INTEGER))
    );

    const queue = [];
    let front = 0;

    // [y, x, cost, direction]
    queue.push([0, 0, 0, 1]);
    queue.push([0, 0, 0, 3]);

    while (front < queue.length) {
        const [y, x, curCost, dir] = queue[front++];

        for (let k = 0; k < 4; k++) {
            const ny = y + dy[k];
            const nx = x + dx[k];
            const nextCost = curCost + (dir === k ? 100 : 600);

            if (
                ny < 0 ||
                nx < 0 ||
                ny >= N ||
                nx >= N ||
                board[ny][nx] === 1 ||
                cost[ny][nx][k] <= nextCost
            ) {
                continue;
            }

            cost[ny][nx][k] = nextCost;
            queue.push([ny, nx, nextCost, k]);
        }
    }

    return Math.min(...cost[N - 1][N - 1]);
}

File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.java
import java.util.*;

public class Solution {
    public int solution(int[][] board) {
        int[] dy = {-1, 1, 0, 0}, dx = {0, 0, -1, 1};
        int N = board.length;

        int[][][] cost = new int[N][N][4];
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                Arrays.fill(cost[i][j], Integer.MAX_VALUE);

        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{0, 0, 0, 1});
        queue.add(new int[]{0, 0, 0, 3});

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();

            for (int k = 0; k < 4; k++) {
                int ny = cur[0] + dy[k], nx = cur[1] + dx[k];
                int c = cur[2] + (cur[3] == k ? 100 : 600);

                if (ny < 0 || nx < 0 || ny >= N || nx >= N || board[ny][nx] == 1 || cost[ny][nx][k] <= c) continue;
                cost[ny][nx][k] = c;
                queue.add(new int[]{ny, nx, c, k});
            }
        }

        return Arrays.stream(cost[N - 1][N - 1]).min().getAsInt();
    }
}